module/zstd/zstd.c

```diff
@@ -0,0 +1,508 @@
+	}
+	if (z == NULL) {
+		return (NULL);
+	}
+
+	z->kmem_magic = ZSTD_KMEM_MAGIC;
+	z->kmem_type = type;
+	z->kmem_size = nbytes;
+
+	return ((void*)z + (sizeof (struct zstd_kmem)));
+}
+
+extern void
+zstd_free(void *opaque __unused, void *ptr)
+{
+	struct zstd_kmem *z = ptr - sizeof (struct zstd_kmem);
+
+	ASSERT3U(z->kmem_magic, ==, ZSTD_KMEM_MAGIC);
+	ASSERT3U(z->kmem_type, <, ZSTD_KMEM_COUNT);
+	ASSERT3U(z->kmem_type, >=, ZSTD_KMEM_UNKNOWN);
+
+	if (z->kmem_type == ZSTD_KMEM_UNKNOWN) {
+		kmem_free(z, z->kmem_size);
+	} else {
+		kmem_cache_free(zstd_kmem_cache[z->kmem_type], z);
+	}
+}
+#ifndef _KERNEL
+#define	__init
+#define	__exit
+#endif
+
+extern int __init
+zstd_init(void)
+{
+	int i;
+
+	/* There is no estimate function for the CCtx itself */
+	zstd_cache_size[1].kmem_magic = ZSTD_KMEM_MAGIC;
+	zstd_cache_size[1].kmem_type = 1;
+	zstd_cache_size[1].kmem_size = P2ROUNDUP(zstd_cache_config[1].block_size
+	    + sizeof (struct zstd_kmem), PAGESIZE);
+	zstd_kmem_cache[1] = kmem_cache_create(
+	    zstd_cache_config[1].cache_name, zstd_cache_size[1].kmem_size,
+	    0, NULL, NULL, NULL, NULL, NULL, 0);
+
+	/*
+	 * Estimate the size of the ZSTD CCtx workspace required for each record
+	 * size at each compression level.
+	 */
+	for (i = 2; i < ZSTD_KMEM_DCTX; i++) {
+		ASSERT(zstd_cache_config[i].cache_name != NULL);
+		zstd_cache_size[i].kmem_magic = ZSTD_KMEM_MAGIC;
+		zstd_cache_size[i].kmem_type = i;
+		zstd_cache_size[i].kmem_size = P2ROUNDUP(
+		    ZSTD_estimateCCtxSize_usingCParams(
+		    ZSTD_getCParams(zstd_cache_config[i].compress_level,
+		    zstd_cache_config[i].block_size, 0)) +
+		    sizeof (struct zstd_kmem), PAGESIZE);
+		zstd_kmem_cache[i] = kmem_cache_create(
+		    zstd_cache_config[i].cache_name,
+		    zstd_cache_size[i].kmem_size,
+		    0, NULL, NULL, NULL, NULL, NULL, 0);
+	}
+	/* Estimate the size of the decompression context */
+	zstd_cache_size[i].kmem_magic = ZSTD_KMEM_MAGIC;
+	zstd_cache_size[i].kmem_type = i;
+	zstd_cache_size[i].kmem_size = P2ROUNDUP(ZSTD_estimateDCtxSize() +
+	    sizeof (struct zstd_kmem), PAGESIZE);
+	zstd_kmem_cache[i] = kmem_cache_create(zstd_cache_config[i].cache_name,
+	    zstd_cache_size[i].kmem_size, 0, NULL, NULL, NULL, NULL, NULL, 0);
+
+	/* Sort the kmem caches for later searching */
+	zstd_qsort(zstd_cache_size, ZSTD_KMEM_COUNT, sizeof (struct zstd_kmem),
+	    zstd_compare);
+	return (0);
+}
+
+extern void __exit
+zstd_fini(void)
+{
+	int i, type;
+
+	for (i = 0; i < ZSTD_KMEM_COUNT; i++) {
+		type = zstd_cache_size[i].kmem_type;
+		if (zstd_kmem_cache[type] != NULL) {
+			kmem_cache_destroy(zstd_kmem_cache[type]);
+		}
+	}
+}
+
+
+#if defined(_KERNEL)
+module_init(zstd_init);
+module_exit(zstd_fini);
+EXPORT_SYMBOL(zstd_fini);
+EXPORT_SYMBOL(zstd_init);
+EXPORT_SYMBOL(zstd_compress);
+EXPORT_SYMBOL(zstd_decompress);
+EXPORT_SYMBOL(zstd_getlevel);
+
+MODULE_DESCRIPTION("ZSTD Compression for ZFS");
+MODULE_AUTHOR(ZFS_META_AUTHOR);
+MODULE_LICENSE(ZFS_META_LICENSE);
```

Comment:
`MODULE_LICENSE("Dual BSD/GPL");`

Reply:
just reused yours. will change it to the original license